File: ios/StickerPackRecords.swift
import ExpoModulesCore

struct StickerInput: Record {
  @Field var imageUrl: String = ""
  @Field var emojis: [String] = []
  @Field var accessibilityText: String?
}

struct StickerPackInput: Record {
  @Field var identifier: String = ""
  @Field var name: String = ""
  @Field var publisher: String = ""
  @Field var trayImageUrl: String = ""
  @Field var publisherEmail: String?
  @Field var publisherWebsite: String?
  @Field var privacyPolicyWebsite: String?
  @Field var licenseAgreementWebsite: String?
  @Field var iosAppStoreLink: String?
  @Field var androidPlayStoreLink: String?
  @Field var animated: Bool = false
  @Field var imageDataVersion: String = "1"
  @Field var stickers: [StickerInput] = []
}

/// Limits published by WhatsApp: https://github.com/WhatsApp/stickers/tree/main/iOS
enum StickerLimits {
  static let stickerSide = 512
  static let traySide = 96
  static let maxStaticStickerBytes = 100 * 1024
  static let maxAnimatedStickerBytes = 500 * 1024
  static let maxTrayBytes = 50 * 1024
  static let minStickers = 3
  static let maxStickers = 30
  static let maxEmojis = 3
  static let maxChars = 128
  static let maxStaticAccessibilityChars = 125
  static let maxAnimatedAccessibilityChars = 255
}

final class StickerPackException: GenericException<String> {
  override var code: String { "ERR_WHATSAPP_STICKERS" }
  override var reason: String { param }
}

File: android/src/main/java/expo/modules/whatsappstickers/StickerPackRecords.kt
package expo.modules.whatsappstickers

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class StickerInput : Record {
  @Field val imageUrl: String = ""
  @Field val emojis: List<String> = emptyList()
  @Field val accessibilityText: String? = null
}

class StickerPackInput : Record {
  @Field val identifier: String = ""
  @Field val name: String = ""
  @Field val publisher: String = ""
  @Field val trayImageUrl: String = ""
  @Field val publisherEmail: String? = null
  @Field val publisherWebsite: String? = null
  @Field val privacyPolicyWebsite: String? = null
  @Field val licenseAgreementWebsite: String? = null
  @Field val iosAppStoreLink: String? = null
  @Field val androidPlayStoreLink: String? = null
  @Field val animated: Boolean = false
  @Field val imageDataVersion: String = "1"
  @Field val stickers: List<StickerInput> = emptyList()
}

/** Limits published by WhatsApp: https://github.com/WhatsApp/stickers/tree/main/Android */
object StickerLimits {
  const val STICKER_SIDE = 512
  const val TRAY_SIDE = 96
  const val MAX_STATIC_STICKER_BYTES = 100 * 1024
  const val MAX_ANIMATED_STICKER_BYTES = 500 * 1024
  const val MAX_TRAY_BYTES = 50 * 1024
  const val MIN_STICKERS = 3
  const val MAX_STICKERS = 30
  const val MAX_EMOJIS = 3
  const val MAX_CHARS = 128
  const val MAX_STATIC_ACCESSIBILITY_CHARS = 125
  const val MAX_ANIMATED_ACCESSIBILITY_CHARS = 255
}

class StickerPackException(message: String) : CodedException("ERR_WHATSAPP_STICKERS", message, null)
